codec-redis/src/main/java/io/netty/handler/codec/redis/RedisEncoder.java

```diff
@@ -0,0 +1,201 @@
+     */
+    public RedisEncoder() {
+        this(FixedRedisMessagePool.INSTANCE);
+    }
+
+    /**
+     * Creates a new instance.
+     * @param messagePool the predefined message pool.
+     */
+    public RedisEncoder(RedisMessagePool messagePool) {
+        this.messagePool = messagePool;
+    }
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, RedisMessage msg, List<Object> out) throws Exception {
+        try {
+            writeRedisMessage(ctx.alloc(), msg, out);
+        } catch (CodecException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new CodecException(e);
+        }
+    }
+
+    private void writeRedisMessage(ByteBufAllocator allocator, RedisMessage msg, List<Object> out) {
+        if (msg instanceof SimpleStringRedisMessage) {
+            writeSimpleStringMessage(allocator, (SimpleStringRedisMessage) msg, out);
+        } else if (msg instanceof ErrorRedisMessage) {
+            writeErrorMessage(allocator, (ErrorRedisMessage) msg, out);
+        } else if (msg instanceof IntegerRedisMessage) {
+            writeIntegerMessage(allocator, (IntegerRedisMessage) msg, out);
+        } else if (msg instanceof FullBulkStringRedisMessage) {
+            writeFullBulkStringMessage(allocator, (FullBulkStringRedisMessage) msg, out);
+        } else if (msg instanceof BulkStringRedisContent) {
+            writeBulkStringContent((BulkStringRedisContent) msg, out);
+        } else if (msg instanceof BulkStringHeaderRedisMessage) {
+            writeBulkStringHeader(allocator, (BulkStringHeaderRedisMessage) msg, out);
+        } else if (msg instanceof ArrayHeaderRedisMessage) {
+            writeArrayHeader(allocator, (ArrayHeaderRedisMessage) msg, out);
+        } else if (msg instanceof ArrayRedisMessage) {
+            writeArrayMessage(allocator, (ArrayRedisMessage) msg, out);
+        } else {
+            throw new CodecException("bad message: " + msg);
+        }
+    }
+
+    private static void writeSimpleStringMessage(ByteBufAllocator allocator, SimpleStringRedisMessage msg,
+                                                 List<Object> out) {
+        writeString(allocator, RedisMessageType.SIMPLE_STRING.value(), msg.content(), out);
+    }
+
+    private static void writeErrorMessage(ByteBufAllocator allocator, ErrorRedisMessage msg,
+                                                 List<Object> out) {
+        writeString(allocator, RedisMessageType.ERROR.value(), msg.content(), out);
+    }
+
+    private static void writeString(ByteBufAllocator allocator, byte type, String message, List<Object> out) {
+        ByteBuf content = ByteBufUtil.encodeString(allocator, CharBuffer.wrap(message), CharsetUtil.UTF_8);
+        ByteBuf buf = allocator.ioBuffer(RedisConstants.TYPE_LENGTH + content.readableBytes() +
+                                         RedisConstants.EOL_LENGTH);
+        buf.writeByte(type);
+        buf.writeBytes(content);
```

Comment:
We can save a copy operation of the string contents if we use `ByteBufUtil.writeUtf8`. We can make [ByteBufUtil.MAX_BYTES_PER_CHAR_UTF8](https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/ByteBufUtil.java#L63) public so we don't have to use a magic number...

```java
ByteBuf buf = allocator.ioBuffer(RedisConstants.TYPE_LENGTH + message.length() * MAX_BYTES_PER_CHAR_UTF8 + RedisConstants.EOL_LENGTH);
buf.writeByte(tyep);
ByteBufUtil.writeUtf8(buf, message);
buf.writeShort(RedisConstants.EOL_SHORT);
out.add(buf);
```
```